internal/kaas/cluster.go

```diff
@@ -0,0 +1,166 @@
+	ApiEndpoint              string
+	ControlPlaneEndpointHost string
+	ControlPlaneEndpointPort int32
+	Region                   string
+	ClusterGroup             string
+	Environment              string
+	CIDR                     []string
+	ControlPlane             *ClusterControlPlane
+	Infrastructure           *ClusterInfrastructure
+}
+
+func GetCluster(k *k8s.Kubernetes, name string) (*Cluster, error) {
+
+	clusterAPICR, err := k.GetCluster(name)
+	if err != nil {
+		clientErr, ok := err.(*clientError.ClientError)
+		if !ok {
+			return nil, clientError.NewClientError(clientErr, clientError.UnexpectedError, fmt.Sprintf("Something went wrong while getting cluster %s", name))
+		} else {
+			if clientErr.ErrorMessage == clientError.ResourceNotFound {
+				return nil, clientError.NewClientError(clientErr, clientError.ResourceNotFound, fmt.Sprintf("Could not find cluster %s", name))
+			} else {
+				return nil, clientError.NewClientError(clientErr, clientError.UnexpectedError, fmt.Sprintf("Error getting cluster %s", name))
+			}
+		}
+	}
+
+	err = ValidateClusterComponents(clusterAPICR)
+	if err != nil {
+		return nil, clientError.NewClientError(err, clientError.InvalidConfiguration, fmt.Sprintf("Cluster %s have an invalid configuration", name))
+	}
+
+	cluster := &Cluster{}
+	err = cluster.GetClusterProperties(clusterAPICR)
+	if err != nil {
+		clientErr, ok := err.(*clientError.ClientError)
+		if !ok {
+			return nil, clientError.NewClientError(err, clientError.UnexpectedError, fmt.Sprintf("An Unexpected error happened while reading cluster %s properties", name))
+		} else {
+			if clientErr.ErrorMessage == clientError.InvalidConfiguration {
+				return nil, clientError.NewClientError(clientErr, clientError.InvalidConfiguration, fmt.Sprintf("Cluster %s is invalid due to missing or invalid labels", name))
+			}
+		}
+	}
+
+	return cluster, nil
+}
+
+func ListClusters(k *k8s.Kubernetes) ([]*Cluster, error) {
+
+	var clusterList []*Cluster
+
+	clusterListAPICR, err := k.ListClusters()
+	if err != nil {
+		clientErr, ok := err.(*clientError.ClientError)
+		if !ok {
+			return nil, clientError.NewClientError(clientErr, clientError.UnexpectedError, "Error listing clusters")
+		} else {
+			if clientErr.ErrorMessage == clientError.ResourceNotFound {
+				return nil, clientErr
+			} else if clientErr.ErrorMessage == clientError.EmptyResponse {
+				return nil, clientErr
+			} else {
+				return nil, clientError.NewClientError(clientErr, clientError.UnexpectedError, "Something went wrong when listing clusters")
+			}
+		}
+	}
+
+	for _, clusterAPICR := range clusterListAPICR.Items {
+		cluster := &Cluster{}
+		err = ValidateClusterComponents(&clusterAPICR)
+		if err != nil {
+			log.Printf("Skipping cluster %s because of invalid configuration: %s", cluster.Name, err.Error())
+			continue
+		}
+		err = cluster.GetClusterProperties(&clusterAPICR)
+		if err != nil {
+			clientErr, ok := err.(*clientError.ClientError)
+			if !ok {
```

Comment:
Maybe this `if` could check for "both" errors (errors for reading a cluster and for invalid or missing labels).

Reply:
You mean ValidateClusterComponents and GetClusterProperties inside the for?

Reply:
As discussed, right now GetClusterProperties needs the validation to be successful, so it's not possible to run both and return the result at once to the user.

But we agreed on better validating the components as having those validations when trying to get them from the ClusterCR, so it was added as a TODO comment.